src/control/lib/spdk/spdk.go

```diff
@@ -41,74 +41,102 @@ import "C"
 import (
 	"fmt"
 	"unsafe"
+
+	"github.com/pkg/errors"
 )
 
-// ENV is the interface that provides SPDK environment management.
-type ENV interface {
-	InitSPDKEnv(int) error
+// Env is the interface that provides SPDK environment management.
+type Env interface {
+	InitSPDKEnv(EnvOptions) error
 }
 
-// Env is a simple ENV implementation.
-type Env struct{}
+// EnvImpl is a simple SPDKEnv implementation.
+type EnvImpl struct{}
 
-// Rc2err returns an failure if rc != 0.
-//
-// TODO: If err is already set then it is wrapped,
-// otherwise it is ignored. e.g.
-// func Rc2err(label string, rc C.int, err error) error {
+// Rc2err returns error from label and rc.
 func Rc2err(label string, rc C.int) error {
-	if rc != 0 {
-		if rc < 0 {
-			rc = -rc
-		}
-		// e := errors.Error(rc)
-		return fmt.Errorf("%s: %d", label, rc) // e
-	}
-	return nil
+	return fmt.Errorf("%s: %d", label, rc)
 }
 
-// InitSPDKEnv initializes the SPDK environment.
-//
-// SPDK relies on an abstraction around the local environment
-// named env that handles memory allocation and PCI device operations.
-// The library must be initialized first.
-func (e *Env) InitSPDKEnv(shmID int, pciWhiteList []string) (err error) {
-	opts := &C.struct_spdk_env_opts{}
+type EnvOptions struct {
+	ShmID        int      // shared memory segment identifier for SPDK IPC
+	MemSize      int      // size in MiB to be allocated to SPDK proc
+	PciWhiteList []string // restrict SPDK device access
+	EnableVMD    bool     // VMD devices should be included
+}
+
+func (o EnvOptions) toC() (opts *C.struct_spdk_env_opts, cWhiteListPtr unsafe.Pointer, err error) {
```

Comment:
This will need to be reworked a bit since Makito's patch to support blacklists has landed.

Reply:
the exclude/blacklist has already been applied during prepare so isn't needed here, the only time we need a whitelist when initing spdk env is when we want to limit access to one instance's bdev_list (currently format and io_server start)